src/utils/deep_merge.ts

```diff
@@ -0,0 +1,43 @@
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import objectAssign from "./object_assign";
+
+/**
+ * Check if an item is an object
+ * @param item
+ * @returns {boolean}
+ */
+function isObject(item: unknown) : boolean {
+  return (item !== null
+          && item !== undefined
+          && typeof item === "object"
+          && !Array.isArray(item));
+}
+
+/**
+ * Deeply merge nested objects
+ * @param target
+ * @param sources
+ * @returns output : merged object
+ */
+export default function deepMerge(target: any, ...sources: any[]): any {
```

Comment:
We could do some kind of "DeepPartial" type:
For example (taken from StackOverflow):

```ts
type IDeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
} : T;
```

Then typing this becomes trivial:
```ts
function deepMerge<T>(target: T, ...sources: Array<DeepPartial<T>>): T;
```

```